File: 808. Soup Servings/SoupServings.py
from typing import Dict, Tuple


class Solution:
    
    def calcProb(self, a: int, b: int, memo: Dict[Tuple[int], int]) -> None:
        
        if a == 0 and b == 0:
            return 0.5
        
        if a == 0:
            return 1
            
        if b == 0:
            return 0
        
        state = (a, b)
        if state in memo:
            return memo[state]
        
        # the 4 operations,
        ops = [[100, 0], [75, 25], [50, 50], [25, 75]]
        prob = 0
        for x, y in ops:
            prob += (0.25) * self.calcProb(max(a-x, 0), max(b-y, 0), memo)
            
        memo[state] = prob
        return prob
        
    
    # O(n^2) time for n < 5000 else O(1)
    # O(n) space,
    # Approach: dp top down, math
    def soupServings(self, n: int) -> float:
        return self.calcProb(n, n, {}) if n < 5000 else 1

File: 808. Soup Servings/soup_servings.rs
use std::collections::{VecDeque, HashMap};
use std::cmp::max;

struct Solution {}

impl Solution {
    pub fn calc_prob(a: i32, b: i32, memo: &mut HashMap<(i32, i32), f64>) -> f64 {
        if a == 0 && b == 0 {
            return 0.5;
        }

        if a == 0 {
            return 1.0;
        }

        if b == 0 {
            return 0.0;
        }

        let state = (a, b);
        if memo.contains_key(&state) {
            return *memo.get(&state).unwrap();
        }

        let mut prob = 0.0;
        let reductions = vec![
            (-100, 0), (-75, -25), 
            (-50, -50), (-25, -75)
            ];

        for (a_reduce, b_reduce) in &reductions {
            let a_new = a + a_reduce;
            let b_new = b + b_reduce;
            prob += 0.25 * Solution::calc_prob(max(a_new, 0), max(b_new, 0), memo);
        }
        memo.insert(state, prob);
        prob
    }

    pub fn soup_servings(n: i32) -> f64 {
        if n >= 5000 {
            return 1.0;
        }

        Solution::calc_prob(n, n, &mut HashMap::new())
    }
}